runtime/src/iree/tooling/buffer_view_matchers.c

```diff
@@ -0,0 +1,668 @@
+// iree_hal_buffer_equality_t
+//===----------------------------------------------------------------------===//
+
+static iree_hal_buffer_element_t iree_hal_buffer_element_at(
+    iree_hal_element_type_t element_type, iree_const_byte_span_t elements,
+    iree_host_size_t index) {
+  iree_host_size_t element_size =
+      iree_hal_element_dense_byte_count(element_type);
+  iree_const_byte_span_t element_data = iree_make_const_byte_span(
+      elements.data + index * element_size, element_size);
+  iree_hal_buffer_element_t element = {
+      .type = element_type,
+  };
+  memcpy(element.storage, element_data.data, element_size);
+  return element;
+}
+
+static iree_status_t iree_hal_append_element_string(
+    iree_hal_buffer_element_t value, iree_string_builder_t* builder) {
+  char temp[64];
+  iree_host_size_t temp_length = 0;
+  IREE_RETURN_IF_ERROR(iree_hal_format_element(
+      iree_make_const_byte_span(value.storage,
+                                iree_hal_element_dense_byte_count(value.type)),
+      value.type, sizeof(temp), temp, &temp_length));
+  return iree_string_builder_append_string(
+      builder, iree_make_string_view(temp, temp_length));
+}
+
+static iree_status_t iree_hal_append_element_mismatch_string(
+    iree_host_size_t index, iree_hal_buffer_element_t expected_element,
+    iree_hal_buffer_element_t actual_element, iree_string_builder_t* builder) {
+  IREE_RETURN_IF_ERROR(iree_string_builder_append_format(
+      builder, "element at index %" PRIhsz " (", index));
+  IREE_RETURN_IF_ERROR(iree_hal_append_element_string(actual_element, builder));
+  IREE_RETURN_IF_ERROR(iree_string_builder_append_string(
+      builder, IREE_SV(") does not match the expected (")));
+  IREE_RETURN_IF_ERROR(
+      iree_hal_append_element_string(expected_element, builder));
+  IREE_RETURN_IF_ERROR(
+      iree_string_builder_append_string(builder, IREE_SV(")")));
+  return iree_ok_status();
+}
+
+static bool iree_hal_compare_strided_elements_exact(
+    iree_hal_element_type_t element_type, iree_host_size_t element_count,
+    iree_const_byte_span_t expected_elements, iree_host_size_t expected_stride,
+    iree_const_byte_span_t actual_elements, iree_host_size_t actual_stride,
+    iree_host_size_t* out_index) {
+  const iree_host_size_t element_size =
+      iree_hal_element_dense_byte_count(element_type);
+  const uint8_t* expected_ptr = expected_elements.data;
+  const uint8_t* actual_ptr = actual_elements.data;
+  for (iree_host_size_t i = 0; i < element_count; ++i) {
+    int cmp = memcmp(expected_ptr, actual_ptr, element_size);
+    if (cmp != 0) {
+      *out_index = i;
+      return false;
+    }
+    expected_ptr += expected_stride * element_size;
+    actual_ptr += actual_stride * element_size;
+  }
+  return true;
+}
+
+// HACK: this should be placed on the equality struct.
+#define IREE_HAL_F16_THRESHOLD 0.001f
+#define IREE_HAL_F32_THRESHOLD 0.0001f
```

Comment:
```suggestion
#define IREE_HAL_F32_THRESHOLD 0.001f
```
We need to relax this a bit. DeepLab requires a tolerance of 1e-3 between platforms. Also https://github.com/iree-org/iree-samples/blob/main/tflitehub/deeplab_v3_test.py#L15

Reply:
👍  I'll add the flag here right away so we can control it per run

Reply:
I was about to do it in a followup PR, but it's awesome that you can add it here.

Reply:
Added `--expected_f32_threshold=` (and f16/f64 variants) that let you override the defaults from the command line.